File: src.csharp/AlphaTab/Core/EcmaScript/ValueTypeMap.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace AlphaTab.Core.EcmaScript
{
    public class ValueTypeMap<TKey, TValue> : IEnumerable<MapEntry<TKey, TValue>>
        where TValue : struct
    {
        private readonly Dictionary<TKey, TValue> _data;

        public ValueTypeMap()
        {
            _data = new Dictionary<TKey, TValue>();
        }

        public ValueTypeMap(IEnumerable<MapEntry<TKey, TValue>> entries)
        {
            _data = entries.ToDictionary(e => e.Key, e => e.Value);
        }

        public double Size => _data.Count;

        public TValue? Get(TKey key)
        {
            if (_data.TryGetValue(key, out var value))
            {
                return value;
            }

            return null;
        }

        public void Set(TKey key, TValue value)
        {
            _data[key] = value;
        }

        public bool Has(TKey key)
        {
            return _data.ContainsKey(key);
        }

        public void Delete(TKey key)
        {
            _data.Remove(key);
        }

        public void Clear()
        {
            _data.Clear();
        }

        public IEnumerator<MapEntry<TKey, TValue>> GetEnumerator()
        {
            return _data.Select(d => new MapEntry<TKey, TValue>(d.Key, d.Value)).GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public void ForEach(Action<TValue, TKey> callback)
        {
            foreach (var kvp in _data)
            {
                callback(kvp.Value, kvp.Key);
            }
        }
    }
}

File: src.kotlin/alphaTab/src/commonMain/kotlin/alphaTab/core/ecmaScript/ValueTypeMap.kt
package alphaTab.core.ecmaScript

import kotlin.collections.Map

class ValueTypeMap<TKey, TValue> : Iterable<Map.Entry<TKey, TValue>> {
    private val _map: HashMap<TKey, TValue>


    public val size: Double
        get() {
            return _map.size.toDouble()
        }

    public constructor() {
        _map = HashMap()
    }

    public constructor(entries: Iterable<Pair<TKey, TValue>>) {
        _map = HashMap()
        _map.putAll(entries);
    }

    public fun get(key: TKey): TValue? {
        return _map[key]
    }

    public fun set(key: TKey, value: TValue) {
        _map[key] = value
    }

    public fun has(key: TKey): Boolean {
        return _map.containsKey(key)
    }

    public fun delete(key: TKey) {
        _map.remove(key)
    }

    public fun clear() {
        _map.clear()
    }

    override fun iterator(): Iterator<Map.Entry<TKey, TValue>> {
        return _map.iterator();
    }
}
